Sources/SwiftGRPCNIO/HTTP1ToRawGRPCServerCodec.swift

```diff
@@ -27,111 +27,200 @@ public enum RawGRPCServerResponsePart {
-          state = .receivedMessageLength(messageLength)
-
-        case .receivedMessageLength(let messageLength):
-          guard let messageBytes = buffer.readBytes(length: numericCast(messageLength)) else { break }
-
-          //! FIXME: Use a slice of this buffer instead of copying to a new buffer.
-          var responseBuffer = ctx.channel.allocator.buffer(capacity: messageBytes.count)
-          responseBuffer.write(bytes: messageBytes)
-          ctx.fireChannelRead(self.wrapInboundOut(.message(responseBuffer)))
-          //! FIXME: Call buffer.discardReadBytes() here?
-          //! ALTERNATIVE: Check if the buffer has no further data right now, then clear it.
-
-          state = .expectingCompressedFlag
-        }
+    if case .ignore = inboundState { return }
+
+    do {
+      switch self.unwrapInboundIn(data) {
+      case .head(let requestHead):
+        inboundState = try processHead(ctx: ctx, requestHead: requestHead)
+
+      case .body(var body):
+        inboundState = try processBody(ctx: ctx, body: &body)
+
+      case .end(let trailers):
+        inboundState = try processEnd(ctx: ctx, trailers: trailers)
       }
+    } catch {
+      ctx.fireErrorCaught(error)
+      inboundState = .ignore
+    }
+  }
+
+  func processHead(ctx: ChannelHandlerContext, requestHead: HTTPRequestHead) throws -> InboundState {
+    guard case .expectingHeaders = inboundState else {
+      throw StateMachineError("expecteded state .expectingHeaders, got \(inboundState)")
+    }
+
+    ctx.fireChannelRead(self.wrapInboundOut(.head(requestHead)))
 
-    case .end(let trailers):
-      if let trailers = trailers {
-        //! FIXME: Better handle this error.
-        print("unexpected trailers received: \(trailers)")
-        return
+    return .expectingBody(.expectingCompressedFlag)
+  }
+
+  func processBody(ctx: ChannelHandlerContext, body: inout ByteBuffer) throws -> InboundState {
+    guard case .expectingBody(let bodyState) = inboundState else {
+      throw StateMachineError("expecteded state .expectingBody(_), got \(inboundState)")
+    }
+
+    return .expectingBody(try processBodyState(ctx: ctx, initialState: bodyState, messageBuffer: &body))
+  }
+
+  func processBodyState(ctx: ChannelHandlerContext, initialState: InboundState.Body, messageBuffer: inout ByteBuffer) throws -> InboundState.Body {
+    var bodyState = initialState
+
+    // Iterate over all available incoming data, trying to read length-delimited messages.
+    // Each message has the following format:
+    // - 1 byte "compressed" flag (currently always zero, as we do not support for compression)
+    // - 4 byte signed-integer payload length (N)
+    // - N bytes payload (normally a valid wire-format protocol buffer)
+    while true {
+      switch bodyState {
+      case .expectingCompressedFlag:
+        guard let compressionFlag: Int8 = messageBuffer.readInteger() else { return .expectingCompressedFlag }
```

Comment:
Nit: name discrepancy between `compressIONFlag` and `[expecting]CompressEDFlag` — rename either for consistency?